packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js

```diff
@@ -130,23 +130,23 @@ module.exports = class GatsbyThemeComponentShadowingResolverPlugin {
           return false
         }
         const existsDir = dir.map(filepath => path.basename(filepath))
-        const exists =
-          // has extension, will match styles.css;
-
-          // import Thing from 'whatever.tsx'
-          // extensions: [.js, .tsx]
-          // site/src/whatever.tsx site/src/whatever.js.
-
-          //exact matches
-          existsDir.includes(path.basename(possibleComponentPath)) ||
-          // .js matches
-          // styles.css.js
-          // whatever.tsx.js
-          this.extensions.find(ext =>
+
+        const isExactPath = existsDir.includes(
+          path.basename(possibleComponentPath)
+        ) // find if there is an exact path match
+
+        if (!isExactPath) {
+          const matchingExtension = this.extensions.find(ext =>
             existsDir.includes(path.basename(possibleComponentPath) + ext)
-          )
-        return exists
+          ) // find extension
+          return matchingExtension
+            ? possibleComponentPath + matchingExtension
+            : false // if extension matches, create path
+        } else {
+          return possibleComponentPath
+        }
       })
+      .find(path => path != false) // if full path matched, return it
```

Comment:
We can make this filter more succinct if all values are falsey or a matched path, a suggestion to consider:

```suggestion
      .find(Boolean) // if full path matched, return it
```